File: backend/expressjs/src/instrument/instrument.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { randomUUID } from "node:crypto";
import { performance } from "node:perf_hooks";
import type { EventBuffer } from "./buffer.js";
import type { EventRecord, Instrumenter, Span } from "./types.js";

export interface TraceContext {
  traceId: string;
  parentSpanId: string | null;
  userId: string | null;
  buffer: EventBuffer;
  instrumenter: InstrumenterImpl;
}

const traceStore = new AsyncLocalStorage<TraceContext>();

export function getTraceContext(): TraceContext | null {
  return traceStore.getStore() ?? null;
}

export function runWithTraceContext<T>(ctx: TraceContext, fn: () => T): T {
  return traceStore.run(ctx, fn);
}

class SpanImpl implements Span {
  private startTime: number;
  private entity_: string | null = null;
  private recordId_: string | null = null;
  private status_: string | null = null;
  private metadata_: Record<string, any> = {};
  private ended = false;

  constructor(
    public readonly traceId: string,
    public readonly spanId: string,
    private readonly parentSpanId: string | null,
    private readonly source: string,
    private readonly component: string,
    private readonly action: string,
    private readonly userId: string | null,
    private readonly buffer: EventBuffer,
  ) {
    this.startTime = performance.now();
  }

  end(): void {
    if (this.ended) return;
    this.ended = true;
    const durationMs = performance.now() - this.startTime;
    const event: EventRecord = {
      trace_id: this.traceId,
      span_id: this.spanId,
      parent_span_id: this.parentSpanId,
      event_type: "system",
      source: this.source,
      component: this.component,
      action: this.action,
      entity: this.entity_,
      record_id: this.recordId_,
      user_id: this.userId,
      duration_ms: Math.round(durationMs * 100) / 100,
      status: this.status_,
      metadata: Object.keys(this.metadata_).length > 0 ? this.metadata_ : null,
    };
    this.buffer.enqueue(event);
  }

  setStatus(status: string): void {
    this.status_ = status;
  }

  setMetadata(key: string, value: any): void {
    this.metadata_[key] = value;
  }

  setEntity(entity: string, recordId?: string): void {
    this.entity_ = entity;
    if (recordId !== undefined) {
      this.recordId_ = recordId;
    }
  }
}

export class InstrumenterImpl implements Instrumenter {
  constructor(private readonly buffer: EventBuffer) {}

  startSpan(source: string, component: string, action: string): Span {
    const ctx = getTraceContext();
    const traceId = ctx?.traceId ?? randomUUID();
    const spanId = randomUUID();
    const parentSpanId = ctx?.parentSpanId ?? null;
    const userId = ctx?.userId ?? null;

    const span = new SpanImpl(
      traceId,
      spanId,
      parentSpanId,
      source,
      component,
      action,
      userId,
      this.buffer,
    );

    // Update the trace context so that subsequent spans created within
    // this context become children of this span.
    if (ctx) {
      ctx.parentSpanId = spanId;
    }

    return span;
  }

  emitBusinessEvent(
    action: string,
    entity: string,
    recordId: string,
    metadata?: Record<string, any>,
  ): void {
    const ctx = getTraceContext();
    const traceId = ctx?.traceId ?? randomUUID();
    const spanId = randomUUID();
    const parentSpanId = ctx?.parentSpanId ?? null;
    const userId = ctx?.userId ?? null;

    const event: EventRecord = {
      trace_id: traceId,
      span_id: spanId,
      parent_span_id: parentSpanId,
      event_type: "business",
      source: "app",
      component: "engine",
      action,
      entity,
      record_id: recordId,
      user_id: userId,
      duration_ms: null,
      status: null,
      metadata: metadata ?? null,
    };
    this.buffer.enqueue(event);
  }
}

class NoopSpan implements Span {
  get traceId(): string {
    return "";
  }
  get spanId(): string {
    return "";
  }
  end(): void {}
  setStatus(_status: string): void {}
  setMetadata(_key: string, _value: any): void {}
  setEntity(_entity: string, _recordId?: string): void {}
}

class NoopInstrumenter implements Instrumenter {
  startSpan(_source: string, _component: string, _action: string): Span {
    return new NoopSpan();
  }
  emitBusinessEvent(
    _action: string,
    _entity: string,
    _recordId: string,
    _metadata?: Record<string, any>,
  ): void {}
}

const noopInstrumenter = new NoopInstrumenter();

/**
 * Returns the Instrumenter from the current trace context, or a NoopInstrumenter
 * if instrumentation is not active for the current request.
 */
export function getInstrumenter(): Instrumenter {
  const ctx = getTraceContext();
  if (ctx) {
    return ctx.instrumenter;
  }
  return noopInstrumenter;
}

/**
 * Returns the current trace ID from the AsyncLocalStorage context, or null
 * if no trace context is active.
 */
export function getTraceId(): string | null {
  const ctx = getTraceContext();
  return ctx?.traceId ?? null;
}

File: backend/golang/internal/instrument/instrument.go
package instrument

import (
	"context"
	"sync"
	"time"

	"github.com/google/uuid"
)

// Context keys
type ctxKey int

const (
	traceIDKey ctxKey = iota
	parentSpanIDKey
	instrumenterKey
)

type userIDKeyType int

const userIDKey userIDKeyType = 0

// Instrumenter interface defines the tracing API.
type Instrumenter interface {
	StartSpan(ctx context.Context, source, component, action string) (context.Context, Span)
	EmitBusinessEvent(ctx context.Context, action, entity, recordID string, metadata map[string]any)
}

// Span interface represents a timed operation span.
type Span interface {
	End()
	SetStatus(status string)
	SetMetadata(key string, value any)
	SetEntity(entity, recordID string)
	TraceID() string
	SpanID() string
}

// Event represents a row in the _events table.
type Event struct {
	TraceID      string         `json:"trace_id"`
	SpanID       string         `json:"span_id"`
	ParentSpanID *string        `json:"parent_span_id"`
	EventType    string         `json:"event_type"`
	Source       string         `json:"source"`
	Component    string         `json:"component"`
	Action       string         `json:"action"`
	Entity       *string        `json:"entity"`
	RecordID     *string        `json:"record_id"`
	UserID       *string        `json:"user_id"`
	DurationMs   *float64       `json:"duration_ms"`
	Status       *string        `json:"status"`
	Metadata     map[string]any `json:"metadata"`
	CreatedAt    time.Time      `json:"created_at"`
}

// newUUID generates a new UUID v4 string.
func newUUID() string {
	return uuid.New().String()
}

// Context helpers

// WithTraceID sets the trace ID in the context.
func WithTraceID(ctx context.Context, traceID string) context.Context {
	return context.WithValue(ctx, traceIDKey, traceID)
}

// GetTraceID returns the trace ID from the context.
func GetTraceID(ctx context.Context) string {
	if v, ok := ctx.Value(traceIDKey).(string); ok {
		return v
	}
	return ""
}

// WithParentSpanID sets the parent span ID in the context.
func WithParentSpanID(ctx context.Context, spanID string) context.Context {
	return context.WithValue(ctx, parentSpanIDKey, spanID)
}

func getParentSpanID(ctx context.Context) string {
	if v, ok := ctx.Value(parentSpanIDKey).(string); ok {
		return v
	}
	return ""
}

// WithInstrumenter sets the instrumenter in the context.
func WithInstrumenter(ctx context.Context, inst Instrumenter) context.Context {
	return context.WithValue(ctx, instrumenterKey, inst)
}

// GetInstrumenter returns the instrumenter from the context,
// or a NoopInstrumenter if none is set.
func GetInstrumenter(ctx context.Context) Instrumenter {
	if v, ok := ctx.Value(instrumenterKey).(Instrumenter); ok {
		return v
	}
	return &NoopInstrumenter{}
}

// WithUserID sets the user ID in the context for instrumentation.
func WithUserID(ctx context.Context, userID string) context.Context {
	return context.WithValue(ctx, userIDKey, userID)
}

// getUserID extracts user ID from context (set by instrumentation middleware).
func getUserID(ctx context.Context) *string {
	if v, ok := ctx.Value(userIDKey).(string); ok {
		return &v
	}
	return nil
}

// InstrumenterImpl is the real instrumenter that enqueues events to the buffer.
type InstrumenterImpl struct {
	buffer *EventBuffer
}

// NewInstrumenter creates a new InstrumenterImpl backed by the given buffer.
func NewInstrumenter(buffer *EventBuffer) *InstrumenterImpl {
	return &InstrumenterImpl{buffer: buffer}
}

// StartSpan creates a new span and returns the updated context.
func (i *InstrumenterImpl) StartSpan(ctx context.Context, source, component, action string) (context.Context, Span) {
	traceID := GetTraceID(ctx)
	parentSpanID := getParentSpanID(ctx)
	spanID := newUUID()

	span := &SpanImpl{
		traceID:      traceID,
		spanID:       spanID,
		parentSpanID: parentSpanID,
		source:       source,
		component:    component,
		action:       action,
		startTime:    time.Now(),
		metadata:     make(map[string]any),
		buffer:       i.buffer,
		userID:       getUserID(ctx),
	}

	// Update context so child spans reference this span as parent
	ctx = WithParentSpanID(ctx, spanID)
	return ctx, span
}

// EmitBusinessEvent emits a one-shot business event (no duration tracking).
func (i *InstrumenterImpl) EmitBusinessEvent(ctx context.Context, action, entity, recordID string, metadata map[string]any) {
	traceID := GetTraceID(ctx)
	spanID := newUUID()
	parentSpanID := getParentSpanID(ctx)

	event := Event{
		TraceID:   traceID,
		SpanID:    spanID,
		EventType: "business",
		Source:    "business",
		Component: "api",
		Action:    action,
		Metadata:  metadata,
	}
	if parentSpanID != "" {
		event.ParentSpanID = &parentSpanID
	}
	if entity != "" {
		event.Entity = &entity
	}
	if recordID != "" {
		event.RecordID = &recordID
	}
	if uid := getUserID(ctx); uid != nil {
		event.UserID = uid
	}
	i.buffer.Enqueue(event)
}

// SpanImpl implements the Span interface with timing and metadata.
type SpanImpl struct {
	traceID      string
	spanID       string
	parentSpanID string
	source       string
	component    string
	action       string
	entity       *string
	recordID     *string
	userID       *string
	status       *string
	startTime    time.Time
	metadata     map[string]any
	buffer       *EventBuffer
	mu           sync.Mutex
	ended        bool
}

func (s *SpanImpl) TraceID() string { return s.traceID }
func (s *SpanImpl) SpanID() string  { return s.spanID }

func (s *SpanImpl) SetStatus(status string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.status = &status
}

func (s *SpanImpl) SetMetadata(key string, value any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.metadata == nil {
		s.metadata = make(map[string]any)
	}
	s.metadata[key] = value
}

func (s *SpanImpl) SetEntity(entity, recordID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.entity = &entity
	if recordID != "" {
		s.recordID = &recordID
	}
}

func (s *SpanImpl) End() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.ended {
		return
	}
	s.ended = true

	durationMs := float64(time.Since(s.startTime).Microseconds()) / 1000.0

	event := Event{
		TraceID:    s.traceID,
		SpanID:     s.spanID,
		EventType:  "system",
		Source:     s.source,
		Component:  s.component,
		Action:     s.action,
		Entity:     s.entity,
		RecordID:   s.recordID,
		UserID:     s.userID,
		DurationMs: &durationMs,
		Status:     s.status,
		Metadata:   s.metadata,
	}
	if s.parentSpanID != "" {
		event.ParentSpanID = &s.parentSpanID
	}
	s.buffer.Enqueue(event)
}
